src/Storages/MaterializedView/RefreshTask.cpp

```diff
@@ -279,10 +279,24 @@ void RefreshTask::wait()
     };
 
     std::unique_lock lock(mutex);
-    refresh_cv.wait(lock, [&] {
-        return state != RefreshState::Running && state != RefreshState::Scheduling &&
-            state != RefreshState::RunningOnAnotherReplica && !scheduling.out_of_schedule_refresh_requested;
-    });
+    /// Waiting for the refresh
+    {
+        /// Note we will wait even in case of WaitingForDependencies
+        if (state == RefreshState::Disabled)
+            return;
+
+        /// First wait for status change (wait() with predicate could fire before
+        /// the status change, i.e. even when refreshTask() is not even started)
```

Comment:
This doesn't seem right in two ways:
1. This introduces the opposite race condition. E.g. if you do `SYSTEM REFRESH VIEW a; SYSTEM WAIT VIEW a;`, the refresh might complete before the wait begins, and the wait will get stuck waiting for the next refresh (which breaks tests with fake time - next refresh won't start until the test advances the clock).
2. `refresh_cv.wait` may do a spurious wakeup.

Instead of this, I think we should set `state` to `RefreshState::Scheduling` whenever `refresh_task->schedule()` is called. ALTER does such call, so the next `wait` will wait for refreshTask to run and update next_refresh_time.

Reply:
>E.g. if you do SYSTEM REFRESH VIEW a; SYSTEM WAIT VIEW a;, the refresh might complete before the wait begins, and the wait will get stuck waiting for the next refresh (which breaks tests with fake time - next refresh won't start until the test advances the clock)

Nice catch (egh, what a dummy mistake)

>Instead of this, I think we should set state to RefreshState::Scheduling whenever refresh_task->schedule() is called. ALTER does such call, so the next wait will wait for refreshTask to run and update next_refresh_time.

Done.